stdlib/public/core/Vector.swift

```diff
@@ -0,0 +1,649 @@
+  ///                    iteration of the closure initializing the vector.
+  /// - Parameter next: A closure that passes in an inout reference to the
+  ///                   user given mutable state which returns an owned
+  ///                   `Element` instance to insert into the vector.
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  public init<State: ~Copyable, E: Error>(
+    expand state: consuming State,
+    with next: (inout State) throws(E) -> Element
+  ) throws(E) {
+    self = try Builtin.emplace { (rawPtr) throws(E) -> () in
+      let buffer = Vector<count, Element>._initializationBuffer(start: rawPtr)
+
+      for i in 0 ..< count {
+        do throws(E) {
+          try buffer.initializeElement(at: i, to: next(&state))
+        } catch {
+          // The closure threw an error. We need to deinitialize every element
+          // we've initialized up to this point.
+          for j in 0 ..< i {
+            buffer.deinitializeElement(at: j)
+          }
+
+          throw error
+        }
+      }
+    }
+  }
+
+  /// Initializes every element in this vector by running the closure with the
+  /// passed in first element.
+  ///
+  /// This will call the closure 'count' times, where 'count' is the static
+  /// count of the vector, to initialize every element by passing the closure
+  /// an immutable borrow reference to the first element given to the
+  /// initializer. The closure is allowed to throw an error at any point during
+  /// initialization at which point the vector will stop initialization,
+  /// deinitialize every currently initialized element, and throw the given
+  /// error back out to the caller.
+  ///
+  /// - Parameter first: The first value to insert into the vector which will be
+  ///                    passed to the closure as a borrow.
+  /// - Parameter next: A closure that passes in an immutable borrow reference
+  ///                   of the given first element of the vector which returns
+  ///                   an owned `Element` instance to insert into the vector.
+  @available(SwiftStdlib 6.1, *)
+  @_alwaysEmitIntoClient
+  public init<E: Error>(
+    unfold first: consuming Element,
+    with next: (borrowing Element) throws(E) -> Element
+  ) throws(E) {
+    // FIXME: We should be able to mark 'Builtin.emplace' as '@once' or something
+    //        to give the compiler enough information to know we will only run
+    //        it once so it can consume the capture. For now, we use an optional
+    //        and take the underlying value within the closure.
+    var o: Element? = first
+
+    self = try Builtin.emplace { (rawPtr) throws(E) -> () in
+      let buffer = Vector<count, Element>._initializationBuffer(start: rawPtr)
+
+      buffer.initializeElement(at: 0, to: o.take()._consumingUncheckedUnwrapped())
+
+      for i in 1 ..< count {
+        do throws(E) {
+          try buffer.initializeElement(at: i, to: next(buffer[0]))
```

Comment:
I believe we'll want to use the previous item here; we should not be repeatedly passing the first item.

```suggestion
          try buffer.initializeElement(at: i, to: next(buffer[i - 1]))
```